src/future/batch_semaphore.rs

```diff
@@ -0,0 +1,602 @@
+    pub fn try_acquire(&self, num_permits: usize) -> Result<(), TryAcquireError> {
+        let mut state = self.state.borrow_mut();
+        state.acquire_permits(num_permits).map_err(|err| {
+            // Conservatively, the requester causally depends on the
+            // last successful acquire.
+            // TODO: This is not precise, but `try_acquire` causal dependency
+            // TODO: is both hard to define, and is most likely not worth the
+            // TODO: effort. The cases where causality would be tracked
+            // TODO: "imprecisely" do not correspond to commonly used sync.
+            // TODO: primitives, such as mutexes, mutexes, or condvars.
+            // TODO: An example would be a counting semaphore used to guard
+            // TODO: access to N homogenous resources (as opposed to FIFO,
+            // TODO: heterogenous resources).
+            // TODO: More precision could be gained by tracking clocks for all
+            // TODO: current permit holders, with a data structure similar to
+            // TODO: `permits_available`.
+            ExecutionState::with(|s| {
+                s.update_clock(&state.permits_available.last_acquire);
+            });
+            err
+        })
+    }
+
+    fn enqueue_waiter(&self, waiter: &Arc<Waiter>) {
+        let mut state = self.state.borrow_mut();
+
+        trace!("enqueuing waiter {:?} for semaphore {:p}", waiter, &self.state);
+        state.waiters.push_back(waiter.clone());
+
+        assert!(!waiter.has_permits.load(Ordering::SeqCst));
+        assert!(!waiter.is_queued.swap(true, Ordering::SeqCst));
+    }
+
+    fn remove_waiter(&self, waiter: &Arc<Waiter>) {
+        let mut state = self.state.borrow_mut();
+
+        trace!(waiters = ?state.waiters, "removing waiter {:?} from semaphore {:p}", waiter, &self.state);
+
+        // sanity checks
+        assert!(!state.closed);
+        assert!(!waiter.has_permits.load(Ordering::SeqCst));
+
+        let index = state
+            .waiters
+            .iter()
+            .position(|x| Arc::ptr_eq(x, waiter))
+            .expect("did not find waiter");
+
+        state.waiters.remove(index).unwrap();
+        assert!(waiter.is_queued.swap(false, Ordering::SeqCst));
+    }
+
+    /// Acquire the specified number of permits (async API)
+    pub fn acquire(&self, num_permits: usize) -> Acquire<'_> {
+        Acquire::new(self, num_permits)
+    }
+
+    /// Acquire the specified number of permits (blocking API)
+    pub fn acquire_blocking(&self, num_permits: usize) -> Result<(), AcquireError> {
+        crate::future::block_on(self.acquire(num_permits))
+    }
+
+    /// Release `num_permits` back to the Semaphore
+    pub fn release(&self, num_permits: usize) {
+        if num_permits == 0 {
+            return;
+        }
+
+        let mut state = self.state.borrow_mut();
+
+        if ExecutionState::should_stop() {
+            return;
+        }
```

Comment:
Nit: ordering of this and `let mut state = self.state.borrow_mut();` should be swapped